### ISSUE TITLE
The php7 feature of using 'self' as return type results in fatal error (on seup:di:compile)

### ISSUE THREAD

**Issue by allanpaiste** [magento/magento2#24475]
_(Re-reporting [an issue](https://github.com/magento/magento2/issues/24443) with even simpler steps)_

This report addresses an issue that has been talked about before, but has not yet received a solution that would be viable in the long run as the focus has been mostly on making sure that Magento2 code does not use certain PHP7 features rather than making Framework fully capable of supporting all that PHP7 has to offer.

Therefore the description has been written in a form where existing code is manipulated with to make sure that there's no way to perceive this as a issue that deserves a solution of [removing the appearance of the problem](https://github.com/magento/magento2/commit/9c9a3fd069493c9e88406a4b6ee79d41835c8d2a) (which is fine as an isolated containment of the issue but I would humbly advise against closing the [ticket](https://github.com/magento/magento2/issues/22769) based on said containment).

Same issue reported in different scenarios in:

https://github.com/magento/magento2/issues/24443
https://github.com/magento/magento2/issues/22769
https://github.com/magento/magento2/issues/23084
https://github.com/magento/magento2/issues/11905

This issue created on purpose to have an open issue on this topic as all that were previously reported have since been closed (with "fixed" flag).

### Environment

1. OS: Ubuntu 18.04.2 LTS (Bionic Beaver)
1. PHP: 7.2.19-0ubuntu0.18.04.1 (Bionic Beaver default); 
   Other 7.2 releases are OK as well
   No other PHP version needs to be tested.
1. Php error reporting level (in /etc/php/7.2/cli/php.ini) configured as follows: 
   `error_reporting = E_ALL & ~E_DEPRECATED & ~E_STRICT`

### Preconditions

_(this guide has been written and tested by following ONLY the steps provided with no manual/custom actions taken)_

1. Use any Magento release (report based on [2.3.2](https://packagist.org/packages/magento/community-edition#2.3.2), which at the time of reporting was the latest release), sample data not needed.
   
   ```shell
   composer create-project magento/community-edition issue-24475
   cd issue-24475
   bin/magento module:enable --all
   ```

### Steps to reproduce

_(All actions performed in terminal, re-creating issue that was hidden away with the change [HERE](https://github.com/magento/magento2/commit/9c9a3fd069493c9e88406a4b6ee79d41835c8d2a))_

1. Modify Magento\Framework\HTTP\PhpEnvironment\Response :: **setHeader** method by adding a 'self' return type to it:

   ```shell
   sed -i \
   's/\(ion setHeader(.*\)/\1: self/g' \
   lib/internal/Magento/Framework/HTTP/PhpEnvironment/Response.php
   ```
1. Trigger code generation with:

   ```shell
   bin/magento setup:di:compile
   ```

### Expected result

1. The code generation command finishes without issues

2. Proxy (_generated/code/Magento/Framework/Webapi/Rest/Response/Proxy.php_) reads: 

```
setHeader(
    $name, 
    $value, 
    $replace = false
) : \Magento\Framework\HTTP\PhpEnvironment\Response
```

### Actual result

1. The code generation command crashes (error peppered with new-lines to make it easier to read)

```
Fatal error: Declaration of 

Magento\Framework\Webapi\Rest\Response\Proxy::setHeader(
    $name, 
    $value, 
    $replace = false
): Magento\Framework\Webapi\Rest\Response 

must be compatible with

Magento\Framework\HTTP\PhpEnvironment\Response::setHeader(
    $name, 
    $value, 
    $replace = false
): Magento\Framework\HTTP\PhpEnvironment\Response 
```

2. Proxy (_generated/code/Magento/Framework/Webapi/Rest/Response/Proxy.php_) reads: 

```
setHeader(
    $name, 
    $value, 
    $replace = false
) : \Magento\Framework\Webapi\Rest\Response
```

### Proposed solutions

* (the most awesome one) Support for all PHP7+ features within the code-generator. Proposed fix is available in the comments of an earlier issue [HERE](https://github.com/magento/magento2/issues/22769#issuecomment-499387177).
* Magento Developer Documentation stating that 'self' is not supported and should be avoided at all cost. Providing also an alternative solution (like this one [HERE](https://github.com/magento/magento2/commit/9c9a3fd069493c9e88406a4b6ee79d41835c8d2a))
* Code Generator to give an error that 'self' is not supported and should not be used.

**Comment by allanpaiste** [comment]
If this one also gets closed, then I will give up ;)

**Comment by engcom-Charlie** [comment]
@allanpaiste i confirm this issue, thanks for contributing!

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Charlie 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-19911` were created

**Issue Available**: @engcom-Charlie, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by fsw** [comment]
I have no time to properly test this, but I am 99% sure this would fix this issue:
https://github.com/magento/magento2/issues/22769#issuecomment-499387177

**Comment by magento-engcom-team** [comment]
Hi @allanpaiste. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-24062` was closed as `Fixed`. 

The fix will be available with the upcoming `2.4.3` release.

**Comment by engcom-Kilo** [comment]
Hello @allanpaiste 
Thanks for report. This issue has been fixed in https://github.com/magento-tsg/magento2/commit/f374b9eb9271502adecb6cabafb1a572cab31526
